domain/network/service/space.go

```diff
@@ -169,3 +170,26 @@ func generateFanSubnetID(subnetNetwork, providerID string) string {
 	subnetWithDashes := strings.Replace(strings.Replace(subnetNetwork, ".", "-", -1), "/", "-", -1)
 	return fmt.Sprintf("%s-%s-%s", providerID, network.InFan, subnetWithDashes)
 }
+
+// Provider is the interface that the network service requires to be able to
+// interact with the underlying provider.
+type Provider interface {
+	environs.Networking
+}
+
+// ProviderService provides the API for working with network spaces.
+type ProviderService struct {
+	Service
+	provider func(context.Context) (Provider, error)
+}
+
+// NewProviderService returns a new service reference wrapping the input state.
+func NewProviderService(st State, provider providertracker.ProviderGetter[Provider], logger Logger) *ProviderService {
+	return &ProviderService{
+		Service: Service{
+			st:     st,
+			logger: logger,
+		},
+		provider: provider,
```

Comment:
This means that now we'll have the service layer split between `Service` and `ProviderService`. Can't we have the provider as part of the `Service` (with a check for nil if used in a method)? that way we have everything under the same struct

Reply:
Correct. This is on purpose. We have the same split between `Service` and `WatchableService`. This is to give guard rails for things like model migrations, which don't require watchers. I want people to think __really__ hard when adding model migrations to understand if they also need a provider. I hope they don't, it should reduce the complexity of model migrations and reduce the exposed testing. If they do require a provider then so be it, but it's a conversation they've at least had with themselves.